### PR TITLE
Aspect ratio in crop view - Replace radio buttons with drop down list

### PR DESCRIPTION
**Pull request by schnuti**
Pull Request for Issue #502, #499 

@dgt41 
> For the other issue let’s add a first option: original aspect ratio which should be selected by default

### Summary of Changes
- Replaced the radio buttons with a drop down.
- Added a default to use original image aspect ratio
- Added Landscape and Portrait orientation alternatives
- Changed the Reset to read "No aspect ratio" for cropping without restrictions.

### Testing Instructions
Click on the different options in the drop down and use the Ratio aspects to crop the images.

Please check the Javascript. I'm not that fit and the Dom results has changed in 4.0

Image 1: The field, Image 2 Drop down opened. 
aspect-ratio_1
aspect-ratio_2

### CODE REVIEW

**Comment by brianteeman** [comment]
No Aspect Ratio doesnt really make sense to me. Just "none" should do

Why do the two disabled options have a value?

File: media/plg_media-action_crop/js/crop.js [hunk 1]
```diff
@@ -57,7 +57,16 @@ Joomla.MediaManager.Edit = Joomla.MediaManager.Edit || {};
 			Joomla.MediaManager.Edit.crop.cropper.setData({height: parseInt(this.value)});
 		});
 
-		var elements = document.querySelectorAll("#jform_aspectRatio input");
+		var elements = document.querySelectorAll(".crop-aspect-ratio-option");
```
Comment (dgrammatiko) [inline media/plg_media-action_crop/js/crop.js]: Please make this an array: 
```js
var elements = [].slice.call(document.querySelectorAll(".crop-aspect-ratio-option"));
```

File: media/plg_media-action_crop/js/crop.js [hunk 2]
```diff
@@ -57,7 +57,16 @@ Joomla.MediaManager.Edit = Joomla.MediaManager.Edit || {};
...
+
+		// Set default aspect ratio after numeric check, option has a dummy value
+		var defaultCropFactor = image.naturalWidth / image.naturalHeight;
+		if (!isNaN(+defaultCropFactor) && isFinite(defaultCropFactor)){
+			elements[0].value = defaultCropFactor;
+		}
+			Joomla.MediaManager.Edit.crop.cropper.setAspectRatio(elements[0].value);
+
+		// Set click listener and function to the aspect ratio list
 		for (var i = 0; i < elements.length; i++) {
```
Comment (dgrammatiko) [inline media/plg_media-action_crop/js/crop.js]: and change this for loop to a foreEach:
```js
elements.forEach(function(element) {
  element.addEventListener('click', function (e) {
   Joomla.MediaManager.Edit.crop.cropper.setAspectRatio(this.value);
  });
});
```

**Review by dgrammatiko — CHANGES_REQUESTED** [review]

**Comment by schnuti** [comment]
@brianteeman 
Yes, None would probably be enough. I was only sure that Reset would not be adequate.

I had some minor problems with the list and think different dummy values are needed. i.e. only one "" is allowed, used for the "None" value.

**Comment by schnuti** [comment]
@dgt41 Made an update with the changes. Thanks.

There seem to be a problem with the initialisation. The default is not always active after loading the image for editing. When you click on Reset in the toolbar it becomes active. Any idea? I'll have to debug that.

**Comment by schnuti** [comment]
@dgt41 The problem shouldn't have with this PR to do.
The first load of an image sets wrong pos-x, pos-y, width and height. A reset sets them correct.

After first load
loaded_img

After reset
reset_img

I'll test some more to see if I can find the worm.

**Comment by schnuti** [comment]
@dgt41 I found the problem. Pretty sure. An event is triggered as the Aspect ratio is changed to the new default.
`
Joomla.MediaManager.Edit.crop.cropper.setAspectRatio(elements[0].value);
` 
My guess is that this line executes to early. But move to where?

File: media/plg_media-action_crop/js/crop.js (hunk 2 above)
Comment (schnuti) [reply media/plg_media-action_crop/js/crop.js]: done

File: media/plg_media-action_crop/js/crop.js (hunk 1 above)
Comment (schnuti) [reply media/plg_media-action_crop/js/crop.js]: done

**Comment by schnuti** [comment]
I should mention that the values are show correct in Javascript debugging mode..

**Comment by dgrammatiko** [comment]
probably it has to do with the event listener `change` (change is ok for select elements not for plain inputs), try setting a timeout of 200 ms in those listeners

**Comment by schnuti** [comment]
@dgt41 
I was fishing in the wrong waters. The image takes it's time to load. 
`
var image = document.getElementById('image-preview');
`
How can this be handled? Some kind of "load"?
I set a timeout of 2000 but that can't be the solution. Can it be depending on filesize and web speed?

**Comment by schnuti** [comment]
@dgt41 
I was fishing in the wrong waters. The image takes it's time to load.

`
var image = document.getElementById('image-preview');
`

I added a timeout of 2000 but that can't be the solution. Is the time depending on image size and web speed? Or would 200 be enough? Is there some kind of "load" check available.

**Comment by schnuti** [comment]
@dgt41 
Now it works as expected.
Thanks!

**Review by dgrammatiko — APPROVED** [review]

**Comment by laoneo** [comment]
Why is landscape and portrait disabled? If you want to make them grouping, probably you should use the group form field then.

**Comment by schnuti** [comment]
@laoneo Sorry, wasn't aware of grouped list. I'll have to check some of my very old code where I probably should change to grouped list.
PR updated.

**Comment by brianteeman** [comment]
Is anything supposed to happen when I select a ratio?
When I change a dimension then something happens but not when I select a ratio. See movie below

mm

**Comment by schnuti** [comment]
Changing the aspect ratio should change the chosen area, though with "reset" values. A lot more logic would be needed to keep one of the values entered. Is the Javascript missing for you? Did it work before the drop down?
The movie I made became too large. I'll see if I can reduce the size.

**Comment by schnuti** [comment]
The Movie. The first width input changes the height according to the "default" aspect ratio. i.e. the loaded images width and height ratio.

aspect_ratio

**Comment by brianteeman** [comment]
It was a completely clean install from your branch

**Comment by schnuti** [comment]
Where you have included this PR? 
If nothing happens when you chose another aspect ratio I suppose the crop.js file isn't working as by me.
Some other tester?

**Comment by schnuti** [comment]
To check if something might be missing I made a clean install from my own patch-2 branch including this patch.
Everything works fine and as expected.

**Comment by schnuti** [comment]
@brianteeman 
A closer look! You are probably using the first version of this PR. I can see, that you don't have the grouped list. Before that, a Javascript problem was discussed above and resolved. You'd better use the latest version.
(For a quick test with the old version, you can click once on the reset button in the toolbar. This should activate the aspect ratio options. )

**Comment by brianteeman** [comment]
This PR is your branch - patch-2
That is what I tested

**Comment by schnuti** [comment]
@brianteeman  Sorry for almost not believing you. I think I found out that some code is browser dependent. I'll have to give another developer tool a try, favourite is Firefox.

Not a defence - but it looks like IE11 (latest automatically updated version on my Win10) is not supported at all by MM.

@dgt41 Do you have an idea what could be browser dependent in the code?
image.addEventListener('load' ? The aspect ratio options do obviously not get initialised correct.

**Comment by brianteeman** [comment]
I can confirm it works with firefox but not with chrome

**Comment by dgrammatiko** [comment]
@schnuti I have no clue why chrome is failing here, try to debug `console.log(image.naturalWidth);` and similarly for the naturalHeight. Is it 0? try maybe `parseInt(image.naturalWidth)` 

About IE11, probably we're missing couple polyfills there (for sure the promise one)

**Comment by schnuti** [comment]
@dgt41 
The code is going through correct but the events for the options are not "active" in Chrome.

`
    // Set click listener and function to the aspect ratio list

    elements.forEach(function(element) {
        element.addEventListener('click', function (e) {
             Joomla.MediaManager.Edit.crop.cropper.setAspectRatio(this.value);
` 

The line addEventListener runs for all option elements in Chrome **and** Firefox.
The click event setAspectRatio only in FF.
My guess is that the option elements are tricky. The old replaced radio buttons are ok in both Chrome and FF. 
A bit desperate I tested the old code with no better result

`
for (var i = 0; i < elements.length; i++) {
`

**Comment by schnuti** [comment]
@dgt41  It looks as I was right. Chrome and Safari ignores click events on select options.
What do you say about those changes? Tested with FF and Chrome.
1. Catch the change event on the parent (select)

```
    document.getElementById('jform_aspectRatio').addEventListener("change", function (e) {
        Joomla.MediaManager.Edit.crop.cropper.setAspectRatio(this.value);
    });
```

2. Change the default option value

```
    // Wait for the image to load its data
    image.addEventListener('load', function() {
        
        // Get all option elements if future need
        var elements = [].slice.call(document.querySelectorAll(".crop-aspect-ratio-option"));
       
        // Set default aspect ratio after numeric check, option has a dummy value
        var defaultCropFactor = image.naturalWidth / image.naturalHeight;
           
        if (!isNaN(+defaultCropFactor) && isFinite(defaultCropFactor)){
            elements[0].value = defaultCropFactor;
        }
        Joomla.MediaManager.Edit.crop.cropper.setAspectRatio(elements[0].value);
    }); 
```

**Comment by laoneo** [comment]
@schnuti can you commit your fix that we can merge it?

**Comment by schnuti** [comment]
OK! Fix commited.

**Comment by schnuti** [comment]
A tip if you like to use the keyboard. Use "tab" to go from the input fields to "Default aspect ratio". Use the up-, down-, left-, right-keys to change the aspect ratio. An effect of the last change.

**Pull request merged**